src/Controls/src/Core/Platform/Android/MultiPageFragmentStateAdapter.cs

```diff
@@ -30,7 +34,125 @@ public override Fragment CreateFragment(int position)
 
 		public override long GetItemId(int position)
 		{
-			return _page.Children[position].GetHashCode();
+			return GetItemIdByPosition(position).ItemId;
+		}
+
+		public override bool ContainsItem(long itemId)
+		{
+			return GetItemByItemId(itemId) != null;
+		}
+
+		ItemKey GetItemIdByPosition(int position)
+		{
+			CheckItemKeys();
+			var page = _page.Children[position];
+			for (var i = 0; i < keys.Count; i++)
+			{
+				ItemKey item = keys[i];
+				if (item.Page == page)
+				{
+					return item;
+				}
+			}
+
+			ItemKey itemKey = new ItemKey(page, (ik) => keys.Remove(ik));
+			keys.Add(itemKey);
+
+			return itemKey;
+		}
+
+		ItemKey? GetItemByItemId(long itemId)
+		{
+			CheckItemKeys();
+			for (var i = 0; i < keys.Count; i++)
+			{
+				ItemKey item = keys[i];
+				if (item.ItemId == itemId)
+				{
+					return item;
+				}
+			}
+
+			return null;
+		}
+
+		void CheckItemKeys()
+		{
+			for (var i = 0; i < keys.Count; i++)
+			{
+				ItemKey item = keys[i];
+
+				if (!_page.Children.Contains(item.Page))
+				{
+					// Disconnect will remove the ItemKey from the keys list
+					item.Disconnect();
```

Comment:
Will this cause an issue with skipping?

For example, assume 3 items in an array:


```cs
var keys = new [] { 1, 2, 3 };
for (var i = 0; i < keys.Count; i++)
{
    WriteLine($"{i} = {keys[i]}");
    if (i == 2)
    {
        keys.RemoveAt(i);
        WriteLine($"remove, {keys.Count} left");
    }
    else
    {
        WriteLine($"no remove, {keys.Count} left");
    }
}
```

The output would be:

```
0 = 1
no remove, 3 left
1 = 2
remove, 2 left
```

The loop will exit as the item was removed, and now i++ == Count.

Might need to go backwards in this loop...

Reply:
🤦fixed